### DISCUSSION TITLE
shutdown mitmdump 7.0.4 by request raise coroutine 'ConnectionHandler.hook_task' was never awaited

### DISCUSSION THREAD

**Discussion by mrsiano** [mitmproxy/mitmproxy#5000] [General]
#### Problem Description
running mitmdump 7.0.4 with custom addon, following the [shutdown](https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/shutdown.py) example,
when it gets the shutdown() call, it raise `coroutine 'ConnectionHandler.hook_task' was never awaited`

#### Steps to reproduce the behavior:
1. run the following script 
```
from mitmproxy import ctx, http


class NetworkDumper:
    def request(self, flow: http.HTTPFlow):
        if "shutdown" in flow.request.url:
            ctx.master.shutdown()


addons = [NetworkDumper()]
```
#### System Information
Mitmproxy: 7.0.4 binary
Python:    3.9.7
OpenSSL:   OpenSSL 1.1.1l  24 Aug 2021
Platform:  macOS-10.15.7-x86_64-i386-64bit

**Comment by mrsiano** [comment]
Similar issue for done(), when hitting Ctrl + C the done is not evaluated 
```
from mitmproxy import ctx


class NetworkDumper:
    def done(self):
        ctx.log.info("Network Dumper Done")


addons = [NetworkDumper()]
```

**Comment by mrsiano** [comment]
@Prinzhorn @mhils @yejianfengblue ping

**Reply by yejianfengblue** [reply]
sorry I'm not a mitmproxy expert. I just happened to answer my own and the only one question.

**Comment by mrsiano** [comment]
this issue solved by shutting down from different thread. as follow:
```
    def shutdown(self):
        time.sleep(1)
        ctx.master.shutdown()

    def request(self, flow: http.HTTPFlow):
            # DO SOMETHING ...
            ctx.log.info("Shutdown the mitmdump process")
            ctx.master.addons.trigger(DoneHook())
            Thread(target=self.shutdown, args=(), daemon=True).start()
            return
```
